Add transit utilities to convert GtfsModel to Feed compatible with roadway network; update documentation

lmz opened the pull request:
This PR introduces comprehensive transit utilities that enable seamless integration of GTFS (General Transit Feed Specification) data with roadway networks. These utilities support creating multimodal transportation networks by mapping transit stops to roadway nodes, generating transit-only links, and converting schedule-based GTFS data to frequency-based representations suitable for travel demand modeling.

### Motivation

The Metropolitan Transportation Commission (MTC) and other agencies need to create integrated multimodal networks that combine roadway and transit data for travel demand modeling. Previously, this required manual processing and custom scripts. This PR provides a robust, reusable solution for:

  - Converting GTFS feeds to network-compatible formats
  - Mapping bus stops to roadway network nodes
  - Creating transit-only infrastructure (stations, access/egress links)
  - Filtering transit data by geographic boundaries
  - Managing transit agency data

Tested via https://github.com/BayAreaMetro/mtc_wrangler/blob/main/create_baseyear_network/create_mtc_network_from_OSM.py

### Key Features Added

  🚌 Transit Stop-to-Roadway Mapping

  Function: `match_bus_stops_to_roadway_nodes()`
  - Uses spatial indexing (BallTree) for efficient nearest-neighbor matching
  - Configurable search radius (default 300 feet/100 meters)
  - Handles both local and express bus stops
  - Preserves stop attributes while updating geometry to snap to road network

🛤️ Shape-Aware Transit Route Path Finding

  Function: `create_bus_routes()`

  - *Shape-aware routing*: Finds shortest paths through road network that follow GTFS shape geometry

  - Robust stop-to-shape matching algorithm that handles:

    - Stops that don't perfectly align with shape points
    - Routes that exit and re-enter service areas
    - Circular/loop routes
    - Missing or incomplete shape data
  - Creates short bus-only links when road network routing is impossible
  - Handles cases where direct paths don't exist
  - Generates accurate route geometries following actual roads
  - Supports route-specific debugging with `trace_shape_ids`
  - Configurable distance tolerance for shape matching

  🚉 Station and Transit Link Generation

  Function: `add_stations_and_links_to_roadway_network()`
  - Creates new nodes for rail/BRT stations
  - Generates access/egress links connecting stations to nearby roads
  - Configurable walk access distance (default 0.5 miles)
  - Assigns unique model_node_ids avoiding conflicts with road network

  📊 GTFS-to-Feed Conversion

  Function: `create_feed_from_gtfs_model()`
  - Converts schedule-based GTFS to frequency-based representation
  - Aggregates trips by pattern and time period
  - Calculates headways using configurable methods (uniform, mean, median)
  - Handles one-time routes with default frequencies
  - Integrates with roadway network for stop locations

### Technical Implementation

#### Dependencies

  - Required: pandas, geopandas, numpy, shapely, networkx
  - Optional: scikit-learn (for spatial indexing - lazy loaded only when transit functions are used)

#### Performance Optimizations

  - Spatial indexing with BallTree for O(log n) stop matching
  - Vectorized operations for coordinate transformations
  - Efficient graph algorithms for pathfinding

#### Error Handling

  - Comprehensive validation of GTFS data
  - Clear error messages for missing dependencies
  - Graceful handling of missing optional GTFS tables
  - Detailed logging for debugging

#### Code Quality

  - Added comprehensive docstrings with parameter descriptions and examples
  - Included type hints for all functions
  - Added constants for magic numbers
  - Implemented lazy loading for optional dependencies
  - Fixed all linting issues (Ruff, MyPy)
  - Added noqa comments where complexity is justified by business logic

#### Testing Considerations

  The transit utilities have been tested with:
  - 511 San Francisco Bay Area GTFS feed (September 2023)
  - Multiple transit agencies (BART, Muni, Caltrain, AC Transit, etc.)
  - Networks ranging from single counties to 9-county regions
  - Various coordinate reference systems

#### Related Files

  - network_wrangler/utils/transit.py - Main transit utilities (2,900+ lines)
  - network_wrangler/transit/io.py - Transit I/O operations
  - network_wrangler/models/gtfs/ - GTFS data models
  - network_wrangler/transit/feed/ - Feed object implementation

#### Breaking Changes

  None. All transit utilities are additive and don't affect existing functionality.

### Notes for Reviewers

  1. Large file: utils/transit.py is necessarily large (2,900+ lines) as it contains comprehensive transit processing logic. Functions are well-organized and documented.
  2. Optional dependency: scikit-learn is only required if transit utilities are used, avoiding unnecessary dependencies for users who only need roadway functionality.
  3. MTC-specific defaults: Some defaults (like 300-foot search radius) are based on MTC standards but are configurable for other regions.
  4. Performance: Processing large GTFS feeds (like the full Bay Area with 20+ agencies) takes 5-10 minutes on typical hardware.
  5. I iterated a lot so there are a lot of useless commits; recommend squashing for merge...

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1211142147352207

lmz committed:
Let's do San Francisco -- it's more interesting

lmz committed:
Merge remote-tracking branch 'upstream/main'

lmz committed:
Merge remote-tracking branch 'upstream/main'

lmz committed:
Documentation updates for mkdocs

lmz committed:
Update so that documentation will link to pandera docs

lmz committed:
Add mermaid diagram for DataFrameModel subclasses

lmz committed:
More documentation updates

Moving API detail to API_*.md

lmz committed:
Add links to organization section

lmz committed:
Add pydantic object inventory

lmz committed:
Break up api.md a bit further

lmz committed:
More documentation updates (no real code changes)

lmz committed:
Add object containment diagram

lmz committed:
Add DBModelMixin Inheritance diagram

lmz committed:
Update load_feed_from_path() to read wrangler_flavored GTFS or just GTFS

lmz committed:
In pure GTFS, frequencies.txt is optional

lmz committed:
Documentation fixes and update

lmz committed:
Fix errors in docstrings

lmz committed:
Documentation updates & fixes

lmz committed:
Respect the info direction for std_out_level

lmz committed:
Updates from running pre-commit

lmz committed:
Add pandas objects.inv

lmz committed:
Add function load_roadway_from_dataframes()

lmz committed:
Add file_mode argument to setup_logging()

lmz committed:
add stub for create_feed_from_gtfs_model()

lmz committed:
Initial implementation of create_feed_from_gtfs_model()

lmz committed:
Pure GTFS requires an agency.txt

lmz committed:
Simplify code; gtfs requires these tables

lmz committed:
Handle optional filtering of service_ids in load_feed_from_path

lmz committed:
Fix join issues from mismatch types by forcing trip_id and stop_id columns to be str

lmz committed:
Let user define time periods and don't assign default frequencies

lmz committed:
Refactor slightly

lmz committed:
Add parameter: default_frequency_for_onetime_route

lmz committed:
pre-commit fixes

lmz committed:
Match non-street stops to nodes on transit links

lmz committed:
Fix error message

lmz committed:
For non-road transit, look at links as well as nodes for node-mapping

lmz committed:
Log more information about route matching failure

lmz committed:
Fix failure messaging for route matching

lmz committed:
Fix to be consistent with code (e.g. convert_stops_to_wrangler_stops())

lmz committed:
Use ASCII-compatible log messages

lmz committed:
Log distance to nearest stop as well

lmz committed:
Add threshold for stop matching and set default to 0.25 miles

lmz committed:
Log agency with stops for match failures

lmz committed:
Bug fixes and simplification

lmz committed:
Bug fix required to use RoadwayNetwork.move_nodes()

lmz committed:
Add filter_transit_by_boundary() and drop_transit_agency()

lmz committed:
Add ferry_only, similar to bus_only and rail_only

lmz committed:
Some refactoring and simplification and error handling

lmz committed:
Add route truncation option for routes partially in boundary

lmz committed:
Create RoadwayNetwork.split_link()

lmz committed:
Fix _update_props_for_common_idx() since the index isn't nec a RangeIndex

lmz committed:
Bug fix

lmz committed:
Create truncate_route_at_stop()

lmz committed:
Fix error message spacing

lmz committed:
Add __repr__() to Feed and GtfsModel

lmz committed:
Fix typo

lmz committed:
Set crs for feed_tables["stops"]

lmz committed:
Fix more error message spacing

lmz committed:
Create match_stops_for_mixed_traffic()

Also trying to simplify code and create fewer copies of things

lmz committed:
Fix so that parent_stations of stops aren't filtered out

lmz committed:
Update filter_transit_by_boundary(),  drop_transit_agency() and truncate_route_at_stop() to operate in place

lmz committed:
Handle parent stations - don't drop those stop_ids

lmz committed:
parent_stations were not classified as stations or mixed-traffic stops; inherit this classification from their children

lmz committed:
Move logging to be inline to match_route_stops_with_connectivity()

lmz committed:
One more log fix

lmz committed:
Use rail_only, bus_only, ferry_only instead of transit

lmz committed:
Add first iteration of create_feed_shapes()

And make crs_units/local_crs parameters

lmz committed:
Fix log message spacing

lmz committed:
Add ferry mode, ferry_only is part of transit

lmz committed:
Fix bugs

lmz committed:
One more bug fix to get the graph to work

lmz committed:
Add model validator, validate_referential_integrity

lmz committed:
Check nodes for links using model_node_id

lmz committed:
In create_feed_shapes(), find intermediate shape points using shortest path

lmz committed:
Fix typo

lmz committed:
Iterate on create_feed_shapes() and move comments to docstrings

lmz committed:
Add special shapes processing for rail

lmz committed:
Update documentation and throw exception for stop sequences fails

lmz committed:
Update split_link() to handle transit links

lmz committed:
Fix bug in match_stops_with_connectivity_for_stations()

which needs to be direction-aware
and iterate on shapes updating

lmz committed:
Use drive graph to identify drive-accessible nodes

lmz committed:
Fix issue preventing debug write of invalid_shape_sequences_gdf

lmz committed:
Add better documentation and rename create_feed_shapes_for_rail

to create_feed_shapes_for_station_only_route_types

lmz committed:
Add skip_stop_agencies parameter to create_feed_shapes_for_station_only_route_types()

lmz committed:
Minor logging updates

lmz committed:
Update write_transit() to handle TransitNetwork, Feed or GtfsModel

lmz committed:
Fix bugs in match_stops_with_connectivity_for_stations()

lmz committed:
Push bug fix for logic to prevent a multiple stations mapping to the same node

lmz committed:
Make sure stations are within max distance

lmz committed:
Simplify logic for stops that repeat and focus shape logging to random_shape_ids

lmz committed:
Add build_shape_links_gdf()

lmz committed:
Create __str__ representation of RoadwayNetwork

lmz committed:
Fix bug in load_roadway_from_dataframes() implementation

lmz committed:
Implement create_feed_frequencies() to use shape_id as representative trip

lmz committed:
Completed add_additional_data_to_shapes()

lmz committed:
Small bug fixes and deleting unused code

lmz committed:
Create transit stop-to-stop links with shapes

lmz committed:
Format error message

lmz committed:
Clarify/simplify debug logging

lmz committed:
Implement add_stations_and_links_to_roadway_network and trace_shape_ids

lmz committed:
Fix warnings

lmz committed:
Fix bug in first stop_id; make non_station_stop_links_df a GeoDataFrame

lmz committed:
Fix bug with creating link if there are no intermediate shape points

lmz committed:
Revert to original max_rows

lmz committed:
Removing deprecated methods; updated match_bus_stops_to_roadway_nodes()

lmz committed:
Starting create_bus_routes()

Deprecate a bunch of unneeded code

lmz committed:
Fix bug with creating stop nodes for stop_ids serving buses and light rail

lmz committed:
Finish implementing create_bus_routes()

lmz committed:
Update docstrings

lmz committed:
Update column set in feed_tables['shapes'] to shape_model_node_id

lmz committed:
Update column name to shape_mode_node_id

lmz committed:
Consolidate stops to stop_id (model node id)

lmz committed:
Deprecate stop_in_mixed_traffic, child_stop_in_mixed_traffic as they're not used

lmz committed:
Deprecate unused method build_shape_links_gdf()

lmz committed:
Set 'highway' value for transit links, if it's a column

lmz committed:
Updates for pre-commit (spacing mostly)

lmz committed:
More fixes for pre-commit

lmz committed:
More pre-commit fixes

lmz committed:
Another pre-commit fix

lmz committed:
Removing unused constants and fix mkdocs formatting

e-lo commented on network_wrangler/utils/transit.py:
```diff
@@ -0,0 +1,3087 @@
+
+    # Build spatial index for bus nodes
+    try:
+        from sklearn.neighbors import BallTree  # noqa: PLC0415
```
Nice! I bet this is a lot faster than the default RTREE that geopandas uses.  Maybe we should create a `netpandas` extension

e-lo commented on network_wrangler/utils/transit.py:
```diff
@@ -0,0 +1,3087 @@
+        f"\n{feed_tables['stops'][['valid_match', f'match_distance_{crs_units}']].describe()}"
+    )
+
+    # Update shapes table similarly
```
Logically, don't you want to update the shapes table first and then do stops which are along shapes?

lmz replied:
Yes! Updated thusly.

e-lo commented on network_wrangler/utils/transit.py:
```diff
@@ -0,0 +1,3087 @@
+                "valid_match",
+            ]
+        ],
+        on="stop_id",
```
I'm not sure I get how shapes table ends up with a good `stop_id`? I might need to reread and figure out...

e-lo commented on network_wrangler/utils/transit.py:
```diff
@@ -0,0 +1,3087 @@
+        f"\n{feed_tables['stops'][['valid_match', f'match_distance_{crs_units}']].describe()}"
+    )
+
+    # Update shapes table similarly
```
I think I'm a little confused here. Is the algorithm only updating the shapes where there is an entry for the stops?   If you want the shape lat/longs to be associated with the actual network geometry  and valid network links, then you'll want to do this: https://github.com/e-lo/gtfs_wrangler/blob/e070c06103cba824248c1d96a6ca1ca9e3d9ccab/pandatransit/transit/edit/align_shapes.py#L127

e-lo replied:
Maybe the shapes should only be updated in route_shapes_between_stops function (currently called create routes)?

lmz replied:
The stops are now added to the shapes via `add_additional_data_to_shapes()`

e-lo commented on network_wrangler/utils/transit.py:
```diff
@@ -0,0 +1,3087 @@
+        feed_tables["stop_times"].drop(columns=["geometry_bus", "_merge"], inplace=True)
+
+
+def create_bus_routes(  # noqa: PLR0912, PLR0915
```
Pr0bably want to name this "route shapes between stops"

e-lo reviewed (changes requested):
Mostly minor comments/reorgs and answering some questions @lmz had.  The utils/transit part though with the routing network is really difficult to parse in my brain though and ideally would be reworked.  I realize it is probably a port of ranch though so I am ok with that being a TODO ...but would like it marked as such.

lmz committed:
Add back other typing (Dict, List, Tuple)

lmz committed:
Fix UP035 `typing.Dict` is deprecated, use `dict` instead

Also remove _skip_stop_agencies

lmz commented on network_wrangler/models/gtfs/tables.py:
```diff
@@ -165,14 +166,14 @@ class WranglerStopsTable(StopsTable):
         coerce=True, nullable=False, unique=True, description="The model_node_id."
     )
-    stop_id_GTFS: Series[str] = Field(
+    gtfs_stop_id: Series[str] = Field(
```
Oh! That sounds good to me! When I was testing, it seemed like gtfs_stop_id was being set instead of this one, so I'll update to use stop_id_GTFS instead. I like the reasoning.

e-lo replied:
ooo - well we should consolidate on one for sure!  So you might need to do a search.

lmz replied:
Fixed: https://github.com/BayAreaMetro/network_wrangler/commit/e11d362d4ccedb396e1382d2bced3d5c5e72fdc7

lmz committed:
Add filter_to_nodes to load_roadway_from_dataframes()

lmz commented on network_wrangler/roadway/io.py:
```diff
@@ -120,6 +122,66 @@ def load_roadway(
+    processed_nodes_df = data_to_nodes_df(nodes_df, config=config)
+
+    WranglerLogger.debug("Processing links_df through data_to_links_df")
+    processed_links_df = data_to_links_df(links_df, nodes_df=processed_nodes_df)
```
Sounds good! Committed: https://github.com/BayAreaMetro/network_wrangler/commit/2ec399841e6a9f64619ac267fee85e169267efa8

lmz committed:
Update comment to be useful

lmz commented on network_wrangler/params.py:
```diff
@@ -14,14 +14,18 @@
 
 DEFAULT_SEARCH_MODES: list = ["drive"]
+"""
+lmz: Why isn't this any?
```
Thanks; that makes sense (and I replaced my comment as such: https://github.com/BayAreaMetro/network_wrangler/commit/0f3075d1980001142ae7e2b15763475094a220eb)

lmz committed:
Use stop_id_GTFS instead of gtfs_stop_id

lmz commented on network_wrangler/transit/feed/feed.py:
```diff
@@ -5,9 +5,13 @@
 from pandera.typing import DataFrame
 
+# Constants
+MAX_AGENCIES_DISPLAY = 3
```
I'm flexible. I think this got created during pre-commit fixes...

lmz committed:
Modify service_ids_filter arg to be a list rather than a dataframe

lmz committed:
shape_id is not always required for GTFS

From https://gtfs.org/documentation/schedule/reference/#tripstxt:
shape_id is Conditionally Required:
- Required if the trip has a continuous pickup or drop-off behavior defined either in routes.txt or in stop_times.txt.
- Optional otherwise.

lmz committed:
Creat filter_feed_by_service_ids()

lmz committed:
Add todo note

lmz commented on network_wrangler/transit/io.py:
```diff
@@ -49,37 +59,216 @@ def load_feed_from_path(
     feed_dfs = {table: _read_table_from_file(table, file) for table, file in feed_files.items()}
 
-    return load_feed_from_dfs(feed_dfs)
+    if service_ids_filter is not None:
```
Updated: https://github.com/BayAreaMetro/network_wrangler/commit/346a8b3a49449e452537e86fc889d0d4517b2648

lmz committed:
Make kwargs passthrough for read functions

lmz commented on network_wrangler/transit/io.py:
```diff
@@ -49,37 +59,216 @@ def load_feed_from_path(
     try:
         if file.suffix in [".csv", ".txt"]:
-            return pd.read_csv(file)
+            return pd.read_csv(file, low_memory=False)
```
Updated: https://github.com/BayAreaMetro/network_wrangler/commit/340d9366253cf249645ec749a6e67d0997fad21c

lmz committed:
Concatenate before data_to_nodes as that does validation

lmz commented on network_wrangler/roadway/network.py:
```diff
@@ -479,12 +510,15 @@ def add_nodes(
-        )
+        WranglerLogger.debug(f"add_nodes() add_nodes_df:\n{add_nodes_df}")
+        concatenated_df = concat_with_attr([self.nodes_df, add_nodes_df], axis=0)
+        self.nodes_df = validate_df_to_model(concatenated_df, RoadNodesTable)
```
Makes sense! https://github.com/BayAreaMetro/network_wrangler/commit/b8bcca00880489c826eb4232c7da25adba237ede

lmz commented on network_wrangler/roadway/network.py:
```diff
@@ -611,6 +645,242 @@ def delete_nodes(
             self.nodes_df, del_node_ids, ignore_missing=selection.ignore_missing
         )
 
+    def split_link(  # noqa: PLR0912, PLR0915
```
Should I move it?  That would involve some refactoring as this also updates the nodes table and I wrote it to edit the RoadwayNetwork in place

e-lo replied:
Reviewing that we do deleting nodes and links similarly so actually I think I'm ok with it in general.  The reference I was using was when we copy and offest links for managed lanes using [`links.create.copy_links()`](https://github.com/network-wrangler/network_wrangler/blob/main/network_wrangler/roadway/links/create.py#L125) and then add the nodes separately.  I am cool either way.

lmz committed:
Move constant lists related to route type categories to gtfs.py

lmz commented on network_wrangler/utils/transit.py:
```diff
@@ -0,0 +1,3087 @@
+from ..transit.feed.feed import Feed
+from .time import time_to_seconds
+
+MIXED_TRAFFIC_ROUTE_TYPES = [
```
Makes sense! Updated: https://github.com/BayAreaMetro/network_wrangler/commit/08370d9379360df83f411599217b50615804510b

lmz committed:
Move constants to params.py

lmz commented on network_wrangler/utils/transit.py:
```diff
@@ -0,0 +1,3087 @@
+"""
+
+
+FEET_PER_MILE = 5280.0
```
Makes sense! Moved: https://github.com/BayAreaMetro/network_wrangler/commit/dc31011db5ebbf714310746fd119a22937c2ea1c

lmz committed:
Remove unused method

lmz committed:
Fix bug where some light rail links weren't getting added

Because their stops were existing roadway nodes

e-lo commented on network_wrangler/models/gtfs/gtfs.py:
```diff
@@ -52,21 +109,99 @@ class GtfsModel(DBModelMixin):
 
     table_names: ClassVar[list[str]] = [
-        "frequencies",
+        "agency",
```
technically isn't agency optional?

e-lo commented on network_wrangler/models/gtfs/gtfs.py:
```diff
@@ -12,6 +14,55 @@
     StopTimesTable,
     TripsTable,
 )
+from .types import RouteType
```
I tend to think that all of these route type categorizations should be configs since it isn't something inherant in the definition of GTFS itself and is an assumption that you are making.  I wont die on that hill though :-)

e-lo commented on network_wrangler/models/gtfs/gtfs.py:
```diff
@@ -52,21 +109,99 @@ class GtfsModel(DBModelMixin):
+        """Return a string representation of the GtfsModel with table summaries."""
+        lines = ["GtfsModel:"]
+
+        # Add agency info if available
```
I would encapsulate all of these in a directly callable method e.g. `summary()` and then have `__repr__()` call that.

e-lo commented on network_wrangler/models/gtfs/gtfs.py:
```diff
@@ -52,21 +109,99 @@ class GtfsModel(DBModelMixin):
+        # Add route type breakdown if routes exist
+        if hasattr(self, "routes") and self.routes is not None and len(self.routes) > 0:
+            route_type_counts = self.routes.route_type.value_counts().sort_index()
+            route_type_names = {
```
Isn't this set somewhere that can be called?

e-lo commented on network_wrangler/roadway/network.py:
```diff
@@ -611,6 +646,242 @@ def delete_nodes(
+        self,
+        A: int,
+        B: int,
+        new_model_node_id: int,
```
can be optional, then use:`roadway.nodes.create.generate_node_ids()`

e-lo commented on network_wrangler/roadway/network.py:
```diff
@@ -611,6 +646,242 @@ def delete_nodes(
+            B: The model_node_id of the end node of the link to split.
+            new_model_node_id: The model_node_id for the new node to be created at the split point.
+            fraction: The fraction along the link's length where the split occurs (0 < fraction < 1).
+            split_reverse_link: If True, also splits the reverse link if it exists. Defaults to True.
```
is there ever a time that we wouldn't do this if a reverse link exists?

e-lo commented on network_wrangler/roadway/network.py:
```diff
@@ -611,6 +646,242 @@ def delete_nodes(
+            A: The model_node_id of the start node of the link to split.
+            B: The model_node_id of the end node of the link to split.
+            new_model_node_id: The model_node_id for the new node to be created at the split point.
+            fraction: The fraction along the link's length where the split occurs (0 < fraction < 1).
```
so .85 == 85% from A to B

e-lo commented on network_wrangler/roadway/network.py:
```diff
@@ -611,6 +646,242 @@ def delete_nodes(
 
+    def split_link(  # noqa: PLR0912, PLR0915
+        self,
+        A: int,
```
Should this be a [seleciton dictionary](https://github.com/network-wrangler/network_wrangler/blob/main/network_wrangler/models/projects/roadway_selection.py#L49) and then just do the selection using those methods? I'm not wedded to that but am just trying to match existing patterns.  Happy either way.

e-lo commented on network_wrangler/roadway/network.py:
```diff
@@ -611,6 +646,242 @@ def delete_nodes(
+        # Copy all columns except geometry, X, Y, and model_node_id
+        for col in self.nodes_df.columns:
+            if (
+                col not in ["model_node_id", "X", "Y", "geometry", "osm_node_id"]
```
Use params in [NodesTable definitions](https://github.com/network-wrangler/network_wrangler/blob/main/network_wrangler/models/roadway/tables.py#L330) for this list of vars.

e-lo commented on network_wrangler/roadway/network.py:
```diff
@@ -611,6 +646,242 @@ def delete_nodes(
+        new_node_df = pd.DataFrame([new_node_data])
+        self.add_nodes(new_node_df)
+
+        # Split the geometry
```
I think this is a good candidate to encapsulate in a [geo util](https://github.com/network-wrangler/network_wrangler/blob/main/network_wrangler/utils/geo.py) method so that this method isn't so long.

e-lo commented on network_wrangler/roadway/network.py:
```diff
@@ -611,6 +646,242 @@ def delete_nodes(
+        # Clear shape_id as we're creating new geometry
+        link2_data["shape_id"] = None
+
+        # Handle reverse link if requested
```
It seems like it would make sense to encapsulate this whole thing in a funct and call it once on forward and once on reverse. As it is it is really verbose.

e-lo reviewed (changes requested):
Thanks for all the edits! A few responses to your responses.

lmz committed:
Fix bug with double bus stops

Add more trace logging

lmz committed:
Rather than filtering stops and putting back parents, don't filter referenced parents to begin with

lmz committed:
ML_access default should be the same as access default

lmz committed:
Initial utility to create centroid connectors

lmz commented:
> @lmz this PR is getting large. Especially for things like "centroid connector utils" where you are adding a specific feature, can we start creating feature branches so we can discuss the merits and implementation of a specific feature?
> 
> IMHO this PR should be "add utils to align most GTFS feeds with roadway networks and network wrangler requirements"

Agree - I'll pull out the centroid connector into another PR, but I needed it in the branch for the MoMo workshop.  I'll move it to another PR post-MOMO. Does that work?  😅

lmz committed:
Revert "Initial utility to create centroid connectors"

This reverts commit 7f4239da3592c326ba70b20fbaef2e3015797eb9.

lmz commented:
My notes re: outstanding issues:

High Priority (Main Blocker)

  1. Routing network code complexity (utils/transit.py) - breaking this up into smaller and more understandable pieces

  Code Organization/Refactoring

  2. Consider moving add_midpoint_nodes to roadway.links.edit domain (you asked if you should move it)
  3. Extract node filtering logic (utils/transit.py:530) to roadway/nodes/filters.py
  4. Encapsulate verbose forward/reverse link code (roadway/network.py:781) into a single reusable function
  5. Encapsulate geometry operations (roadway/network.py:729) in geo util methods

  API/Design Questions

  6. Route type categorizations (models/gtfs/gtfs.py:17) - Should be configurable, not hardcoded
  7. Graph building necessity (utils/transit.py:520) - Evaluate if the graph-based approach is needed for performance
  8. Selection dictionary pattern (roadway/network.py:651) - Use selection dictionary for link selection
  9. Make node_id generation optional (roadway/network.py:653) using generate_node_ids()

  Method Organization

  10. Encapsulate GTFS summary (models/gtfs/gtfs.py:135) - Create a summary() method that __repr__() calls
  11. Use NodesTable params (roadway/network.py:719) for variable lists

  Shapes/Routes Logic Clarification

  12. Shapes table update logic (utils/transit.py:664) - Clarify how shapes table gets stop_id and when shapes should be updated
  13. Rename create_routes to route_shapes_between_stops for clarity
  14. Consider moving shapes updates to only happen in the route_shapes_between_stops function

I'm happy to make issues for these (and work on them or assign to others including folks on my team), or keep the PR open 😄.  Thinking on this more, the first option makes it a little easier to delegate around...

lmz committed:
Access shape_id as column name rather than attribute

lmz committed:
add_shapes(): handle empty self.shape_df

lmz committed:
Add function to add link names to nodes

This is useful for things like snapping bus stops which have names like (Market St & Van Ness Ave)

lmz committed:
Since bus stops may have location information (e.g. Market St & 2nd St), use this for more intelligent node matching

lmz committed:
Add better logging for name-aware bus stop matching

lmz committed:
Add shape-awareness to bus stop routing

lmz committed:
Remove log line

lmz committed:
Bug fix: for station links that can't be added because they already exist, make sure existing link has the appropriate rail or ferry access

lmz committed:
Fix quotes

lmz committed:
Add create_debug_links_for_bad_bus_paths()

And other bug fixes

lmz committed:
For shapes that don't match all stops, insert the stop.

lmz committed:
Update shape to stop matching to be more robust

lmz committed:
Generalize fix for stops that don't match to shape points

lmz committed:
More debugging - and yes, this needs refactoring!

lmz committed:
Fix quotes

lmz committed:
More bug fixes

lmz committed:
More bug fixes

lmz committed:
Iterate on matching stop to shape points algorithm

Including breaking it up

lmz committed:
Bug fix

lmz committed:
Bug fixes

lmz committed:
Map stops with poor score to nearest node outside driving network

This is because sometimes the bus stop is in a parking lot etc

lmz committed:
More bug fixes

lmz committed:
Don't crash if no trace_shape_ids

lmz committed:
Added detailed documentation for create_feed_from_gtfs_model()

lmz committed:
Remove non-ascii characters

lmz committed:
Update bus routing to get as close as we can and create short bus-only link

https://app.asana.com/1/11860278793487/project/15119358130897/task/1211808054719862?focus=true

lmz committed:
Bug fix: don't include stops that are non-bus stations

lmz committed:
Go back to table style docs

lmz committed:
Use existing columns rather than creating new ones

lmz committed:
Link attribute name is roadway, not highway

lmz committed:
Read optional tables as well

lmz committed:
Follow-up fixes to reading optional transit tables

lmz committed:
Bad bus paths are for buses

lmz committed:
Fix bug: default_frequency_for_onetime_route is in minutes

e-lo commented:
@lmz - are you still working on this? it would be great ot not leave it branched for too long lest they grow farther apart!

lmz commented:
> @lmz - are you still working on this? it would be great ot not leave it branched for too long lest they grow farther apart!

I am!  And now I need to cherry pick some other relevant commits into this 🥲

lmz committed:
Simplify __repr__ and move bulk of it to summary()

Consistent with other wrangler classes and following request from @e-lo

lmz committed:
Move some functions from utils\transit.py to transit\filter.py

lmz committed:
Move create_feed_frequencies() to models\gtfs\converters.py

lmz committed:
Add default_node_attribute_dict parameter to create_feed_from_gtfs_model()

lmz committed:
Marking as TODO for possible remove based on #408 comments

lmz committed:
Rename create_bus_routes() -> route_shapes_between_stops()

lmz committed:
Move constants to configs\wrangler.py TransitConfig

lmz committed:
Marking additional transit.py TODOs and constrain frequency_method values

lmz committed:
Constrain frequency_method using pydantic

lmz commented on network_wrangler/utils/transit.py:
```diff
@@ -0,0 +1,3087 @@
+    """
+    WranglerLogger.info("Creating frequencies table from feed stop_times data")
+
+    VALID_FREQUENCY_METHOD = ["uniform_headway", "mean_headway", "median_headway"]
```
Fixed

lmz commented on network_wrangler/utils/transit.py:
```diff
@@ -0,0 +1,3087 @@
+METERS_PER_KILOMETER = 1000.0
+"""Meters to kilometers conversion constant."""
+
+MAX_DISTANCE_STOP = {
```
Fixed in https://github.com/BayAreaMetro/network_wrangler/commit/e68281b379806fe4adb706a40268eceb0854419a

lmz committed:
Merge upstream/main into BayAreaMetro:main for PR #408

lmz committed:
Fix ruff check failures after merging upstream/main

lmz committed:
Fix ruff check and format errors

lmz committed:
Fix PLW0108 (inline lambdas) and RUF059 (unused unpacked var)

lmz committed:
style: Fix ruff formatting (ruff 0.15.x)

lmz committed:
Update ruff pre-commit rev to v0.15.8

lmz committed:
Merge remote-tracking branch 'upstream/main'

lmz committed:
Fix test failures: Literal import, fillna(lambda), ferry_only optional column

lmz committed:
Merge upstream/main into BayAreaMetro:main (pandas 3.0 support)

lmz committed:
Lint fix (long line)

lmz committed:
More lint

lmz committed:
Fix test to exclude bools from numeric cols

lmz committed:
Ruff formatting

lmz commented:
Hi @e-lo -
I think I have merged in everything from network-wrangler:main and either addressed your comments or created issues (you created a number of these as well).  So hopefully we can merge in this monster and I can work to resolve issues. 😁 

I *think* the two PR checks are failing because this PR is from a fork... Should I do future PRs from a branch within the network-wrangler org?

e-lo committed:
Merge branch 'main' into main

e-lo reviewed (approved):
Minor style things that I'm going to make a quick follow up commit for but otherwise LGTM! (for now!)

lmz commented:
Thank you!!! 🙏
